src/python/turicreate/_deps/minimal_import.py

```diff
@@ -0,0 +1,367 @@
+
+    def __getattr__(self, attr):
+        self._load_module()
+        return getattr(self._module, attr)
+
+    def __delattr__(self, attr):
+        self._load_module()
+        return delattr(self._module, attr)
+
+    def __setattr__(self, attr, value):
+        """
+        must filter attr by self._my_attrs_ to register members for self
+        instead of the real module object;
+        """
+        # don't use is_loaded; it will trigger getattr
+        # avoid infinite recursion by self._load_module()
+        # when self._load is not set first
+        if attr in self._my_attrs_:
+            # workaround for the recursive setattr
+            return super(DeferredModuleLoader, self).__setattr__(attr, value)
+        else:
+            self._load_module()
+            setattr(self._module, attr, value)
+
+    def is_loaded(self, no_lock=True):
+        if no_lock:
+            return self._loaded
+        else:
+            with _ImportLockContext():
+                return self._loaded
+
+    def is_model(self):
+        return self._is_model
+
+    def reload(self):
+        # for dev purpose
+        with _ImportLockContext():
+            self._module = None
+            self._loaded = False
+            sys.modules.pop(self._name, None)
+
+    # for python `pickle` (serialization) purpose
+    # in order to please lambda worker
+    def get_module(self):
+        with _ImportLockContext():
+            self._load_module()
+            return self._module
+
+    def _load_module(self):
+        # call module.__init__ after import introspection is done
+        if not self._loaded:
+            with _ImportLockContext():
+                # avoid concurrent loading
+                if not self._loaded:
+                    # if it's imported by other pkg after __init__,
+                    # we don't bother to inject ourselves because
+                    # there's no need to lazy load if it's loaded explicitly.
+                    # e.g.
+                    # import numpy as np
+                    # np = DeferredModuleLoader('numpy')
+                    self._module = sys.modules.get(self._name, None)
+                    if self._module is None:
+                        # make sure we clean it up
+                        sys.modules.pop(self._name, None)
+                        self._module = self._init_func(self._name)
+                        # subsequent imports will use the real object
+                        # import pandas from outside turicreate won't know
+                        # nothing inside
+                        sys.modules[self._name] = self._module
+                    self._loaded = True
+
+    # pickle
+    def __reduce__(self):
+        def _init(*args, **kwargs):
+            return DeferredModuleLoader(*args, **kwargs)
+
+        # be aware, singleton must be set false
+        return (_init, (self._name, self._init_func, False))
+
+
+class DeferredCallableLoader(object):
```

Comment:
How is this getting used?

Reply:
not used. Just leave it here for future use.

Reply:
If it's not being used, let's delete it.

Reply:
removed.